### PR TITLE
Run suggestWidget.showSuggestions and showDetails in an animation frame

### PR DESCRIPTION
**Pull request by Tyriar**
This prevents double rendering, both in the keypress task and the following animation frame

Part of #161622

---

I tested to make sure I didn't break anything by doing:

- Typing to refilter
- Backspace to refilter
- Escape to hide
- ctrl+space to toggle details

Let me know if there's other things I need to test.

Before _when re-filtering existing completion items_:

image

After:

image

These numbers vary quite wildly when measured, but the above shows in particular that the double render no longer occurs.

### CODE REVIEW

File: src/vs/editor/contrib/suggest/browser/suggestWidget.ts [hunk 1]
```diff
@@ -104,6 +104,8 @@ export class SuggestWidget implements IDisposable {
 	private _state: State = State.Hidden;
 	private _isAuto: boolean = false;
 	private _loadingTimeout?: IDisposable;
+	private _pendingLayout?: IDisposable;
+	private _pendingShowDetails?: IDisposable;
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/browser/suggestWidget.ts:108]: Suggestion to use `MutableDisposable` instead

File: src/vs/editor/contrib/suggest/browser/suggestWidget.ts [hunk 2]
```diff
@@ -559,9 +561,13 @@ export class SuggestWidget implements IDisposable {
 			this._onDidSelect.resume();
 		}
 
-		this._layout(this.element.size);
-		// Reset focus border
-		this._details.widget.domNode.classList.remove('focused');
+		this._pendingLayout?.dispose();
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/browser/suggestWidget.ts:564]: The pending layout (and details showing) should also be cancelled when hiding the widgets via `hideWidget` and `toggleDetails`

**Review by jrieken — CHANGES_REQUESTED** [review]
Thank you. There is some more cancelling that needs to happen when hiding the widget/details. 

I do see the improvements 🚀  We prevent the "force layout" which is caused by calling `getScrolledVisiblePosition` during `_layout`. I am a little unsure if this is a band aid or not? I believe the use of RAF nicely aligns us with the editor rendering itself and I take this change but I wonder if `_layout` should be improved and not be triggering this at all

**Comment by Tyriar** [comment]
> I am a little unsure if this is a band aid or not?

I wouldn't call this a band aid, I think what we should aim for is any reading or writing of the DOM is done in the animation frame. That's essentially the measure and modify idea you mentioned before, though I imagine it could be tricky to separate the 2 in some cases.

**Comment by Tyriar** [comment]
@jrieken ready for another review

**Review by jrieken — APPROVED** [review]
Thanks @Tyriar 

Pushed another change to cancel RAFs on dispose (corner case)

**Pull request merged**